File: backend-midway/src/controller/log.controller.ts
import { Controller, Get, Post, Inject, Query, Body, Config } from '@midwayjs/core';
import { ApiTags, ApiOperation, ApiQuery, ApiBody } from '@midwayjs/swagger';
import { ElasticsearchService } from '../service/elasticsearch.service';

@Controller('/log')
@ApiTags(['log'])
export class LogController {
  @Inject()
  elasticsearchService: ElasticsearchService;

  @Config('defaultProjectId')
  defaultProjectId: string;

  @Get('/search')
  @ApiOperation({ summary: '查询监控日志' })
  @ApiQuery({ name: 'userId', required: false, description: '用户编号' })
  @ApiQuery({ name: 'type', required: false, description: '日志类型：error, performance, behavior, api' })
  @ApiQuery({ name: 'keyword', required: false, description: '关键字搜索' })
  @ApiQuery({ name: 'startTime', required: false, description: '开始时间（ISO格式）' })
  @ApiQuery({ name: 'endTime', required: false, description: '结束时间（ISO格式）' })
  @ApiQuery({ name: 'page', required: false, description: '页码，从1开始' })
  @ApiQuery({ name: 'pageSize', required: false, description: '每页数量' })
  async search(@Query() query: any) {
    const {
      userId,
      type,
      keyword,
      startTime,
      endTime,
      page = 1,
      pageSize = 20,
    } = query;

    // 如果没有指定时间范围，默认查询最近7天，截止到当前时间
    let finalStartTime = startTime;
    let finalEndTime = endTime;
    
    if (!finalStartTime && !finalEndTime) {
      const now = new Date();
      finalEndTime = now.toISOString();
      finalStartTime = new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000).toISOString();
    }
    
    // 确保时间格式正确（ISO格式）
    if (finalStartTime && !finalStartTime.includes('T')) {
      finalStartTime = new Date(finalStartTime).toISOString();
    }
    if (finalEndTime && !finalEndTime.includes('T')) {
      finalEndTime = new Date(finalEndTime).toISOString();
    }

    // 调试日志
    console.log('🔍 Log search params:', {
      projectId: this.defaultProjectId,
      userId,
      type,
      keyword,
      startTime: finalStartTime,
      endTime: finalEndTime,
      page: parseInt(page) || 1,
      pageSize: parseInt(pageSize) || 20,
      isConnected: this.elasticsearchService.isConnected(),
    });

    const result = await this.elasticsearchService.searchLogs({
      projectId: this.defaultProjectId,
      userId,
      type,
      keyword,
      startTime: finalStartTime,
      endTime: finalEndTime,
      page: parseInt(page) || 1,
      pageSize: parseInt(pageSize) || 20,
    });

    console.log('📊 Log search result:', {
      total: result.total,
      hitsCount: result.hits.length,
      isConnected: this.elasticsearchService.isConnected(),
    });
    
    // 如果未连接，返回提示信息
    if (!this.elasticsearchService.isConnected()) {
      return {
        success: false,
        message: 'Elasticsearch 未连接，请检查 Elasticsearch 服务是否运行',
        data: {
          list: [],
          total: 0,
          page: parseInt(page) || 1,
          pageSize: parseInt(pageSize) || 20,
        },
      };
    }

    return {
      success: true,
      data: {
        list: result.hits,
        total: result.total,
        page: result.page,
        pageSize: result.pageSize,
      },
    };
  }

  @Post('/test')
  @ApiOperation({ summary: '测试写入日志（用于测试）' })
  @ApiBody({ description: '测试日志数据' })
  async testWrite(@Body() body: any) {
    try {
      await this.elasticsearchService.writeLog({
        projectId: this.defaultProjectId,
        type: body.type || 'error',
        userId: body.userId || 'test-user',
        sessionId: body.sessionId || 'test-session',
        url: body.url || 'http://test.com',
        timestamp: new Date().toISOString(),
        message: body.message || 'Test log message',
        ...body,
      });

      return {
        success: true,
        message: '测试日志已写入',
      };
    } catch (error: any) {
      return {
        success: false,
        message: '写入失败: ' + (error.message || '未知错误'),
      };
    }
  }

  @Get('/stats')
  @ApiOperation({ summary: '获取日志统计信息（用于测试）' })
  async stats() {
    try {
      const client = (this.elasticsearchService as any).client;
      if (!client) {
        return {
          success: false,
          message: 'Elasticsearch 未连接',
        };
      }

      // 检查索引是否存在
      const indexExists = await client.indices.exists({ index: 'monitor-logs' });
      
      if (!indexExists) {
        return {
          success: true,
          data: {
            indexExists: false,
            documentCount: 0,
            message: '索引不存在',
          },
        };
      }

      // 获取文档数量
      const countResult = await client.count({ index: 'monitor-logs' });
      
      // 获取最近一条记录
      const searchResult = await client.search({
        index: 'monitor-logs',
        body: {
          query: { match_all: {} },
          size: 1,
          sort: [{ '@timestamp': { order: 'desc' } }],
        },
      });

      return {
        success: true,
        data: {
          indexExists: true,
          documentCount: countResult.count,
          isConnected: this.elasticsearchService.isConnected(),
          lastRecord: searchResult.hits.hits[0]?._source || null,
        },
      };
    } catch (error: any) {
      return {
        success: false,
        message: '查询失败: ' + (error.message || '未知错误'),
        error: error.message,
      };
    }
  }
}

File: backend-springboot/src/main/java/com/monitor/controller/LogController.java
package com.monitor.controller;

import com.monitor.service.ElasticsearchService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.Instant;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 监控日志控制器
 */
@RestController
@RequestMapping("/log")
@RequiredArgsConstructor
public class LogController {
    
    private final ElasticsearchService elasticsearchService;
    
    @Value("${default-project-id:001}")
    private String defaultProjectId;
    
    @GetMapping("/search")
    public ResponseEntity<Map<String, Object>> search(
            @RequestParam(required = false) String userId,
            @RequestParam(required = false) String type,
            @RequestParam(required = false) String keyword,
            @RequestParam(required = false) String startTime,
            @RequestParam(required = false) String endTime,
            @RequestParam(required = false, defaultValue = "1") Integer page,
            @RequestParam(required = false, defaultValue = "20") Integer pageSize) {
        
        // 如果没有指定时间范围，默认查询最近7天，截止到当前时间
        String finalStartTime = startTime;
        String finalEndTime = endTime;
        
        if (finalStartTime == null && finalEndTime == null) {
            Instant now = Instant.now();
            finalEndTime = now.toString();
            finalStartTime = now.minusSeconds(7 * 24 * 60 * 60).toString();
        }
        
        // 确保时间格式正确（ISO格式）
        if (finalStartTime != null && !finalStartTime.contains("T")) {
            finalStartTime = Instant.parse(finalStartTime).toString();
        }
        if (finalEndTime != null && !finalEndTime.contains("T")) {
            finalEndTime = Instant.parse(finalEndTime).toString();
        }
        
        // 调试日志
        System.out.println("🔍 Log search params: " + Map.of(
            "projectId", defaultProjectId,
            "userId", userId != null ? userId : "",
            "type", type != null ? type : "",
            "keyword", keyword != null ? keyword : "",
            "startTime", finalStartTime != null ? finalStartTime : "",
            "endTime", finalEndTime != null ? finalEndTime : "",
            "page", page,
            "pageSize", pageSize,
            "isConnected", elasticsearchService.isConnected()
        ));
        
        Map<String, Object> params = new HashMap<>();
        params.put("projectId", defaultProjectId);
        params.put("userId", userId);
        params.put("type", type);
        params.put("keyword", keyword);
        params.put("startTime", finalStartTime);
        params.put("endTime", finalEndTime);
        params.put("page", page);
        params.put("pageSize", pageSize);
        
        Map<String, Object> result = elasticsearchService.searchLogs(params);
        
        System.out.println("📊 Log search result: " + Map.of(
            "total", result.get("total"),
            "hitsCount", ((List<?>) result.get("hits")).size(),
            "isConnected", elasticsearchService.isConnected()
        ));
        
        // 如果未连接，返回提示信息
        if (!elasticsearchService.isConnected()) {
            return ResponseEntity.ok(Map.of(
                "success", false,
                "message", "Elasticsearch 未连接，请检查 Elasticsearch 服务是否运行",
                "data", Map.of(
                    "list", List.of(),
                    "total", 0,
                    "page", page,
                    "pageSize", pageSize
                )
            ));
        }
        
        return ResponseEntity.ok(Map.of(
            "success", true,
            "data", Map.of(
                "list", result.get("hits"),
                "total", result.get("total"),
                "page", result.get("page"),
                "pageSize", result.get("pageSize")
            )
        ));
    }
    
    @PostMapping("/test")
    public ResponseEntity<Map<String, Object>> testWrite(@RequestBody Map<String, Object> body) {
        try {
            Map<String, Object> logData = new HashMap<>();
            logData.put("projectId", defaultProjectId);
            logData.put("type", body.getOrDefault("type", "error"));
            logData.put("userId", body.getOrDefault("userId", "test-user"));
            logData.put("sessionId", body.getOrDefault("sessionId", "test-session"));
            logData.put("url", body.getOrDefault("url", "http://test.com"));
            logData.put("timestamp", Instant.now().toString());
            logData.put("message", body.getOrDefault("message", "Test log message"));
            logData.putAll(body);
            
            elasticsearchService.writeLog(logData);
            
            return ResponseEntity.ok(Map.of(
                "success", true,
                "message", "测试日志已写入"
            ));
        } catch (Exception e) {
            return ResponseEntity.ok(Map.of(
                "success", false,
                "message", "写入失败: " + (e.getMessage() != null ? e.getMessage() : "未知错误")
            ));
        }
    }
    
    @GetMapping("/stats")
    public ResponseEntity<Map<String, Object>> stats() {
        // TODO: 实现日志统计信息查询
        return ResponseEntity.ok(Map.of(
            "success", true,
            "data", Map.of(
                "indexExists", false,
                "documentCount", 0,
                "isConnected", elasticsearchService.isConnected(),
                "message", "功能待实现"
            )
        ));
    }
}

